geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/KryoVisibilityRowEncoder.scala

```diff
@@ -26,52 +27,117 @@ import org.opengis.feature.simple.SimpleFeatureType
 class KryoVisibilityRowEncoder extends RowEncodingIterator {
 
   private var sft: SimpleFeatureType = _
-  private var nullBytes: Array[Array[Byte]] = _
-  private var offsets: Array[Int] = _
+  private var count: Int = -1
+  private var attributes: Array[(Array[Byte], Int, Int)] = _ // (bytes, offset, length)
+
+  private var nullBytesV2: Array[Array[Byte]] = _
+  private var offsetsV2: Array[Int] = _
   private var offsetStart: Int = -1
 
   private val output: Output = new Output(128, -1)
 
-  override def init(source: SortedKeyValueIterator[Key, Value],
-                    options: java.util.Map[String, String],
-                    env: IteratorEnvironment): Unit = {
+  override def init(
+      source: SortedKeyValueIterator[Key, Value],
+      options: java.util.Map[String, String],
+      env: IteratorEnvironment): Unit = {
     super.init(source, options, env)
-
     sft = IteratorCache.sft(options.get(KryoVisibilityRowEncoder.SftOpt))
-    if (offsets == null || offsets.length != sft.getAttributeCount) {
-      offsets = Array.ofDim[Int](sft.getAttributeCount)
-    }
-    val cacheKey = CacheKeyGenerator.cacheKey(sft)
-    nullBytes = KryoFeatureSerialization.getWriters(cacheKey, sft).map { writer =>
-      output.clear()
-      writer(output, null)
-      output.toBytes
+    count = sft.getAttributeCount
+    if (attributes == null || attributes.length < count) {
+      attributes = Array.ofDim[(Array[Byte], Int, Int)](count)
     }
+    nullBytesV2 = null // lazily initialized when we hit v2 encoded data
```

Comment:
Doesn't line 33 set this to null already?  Or is the _ an empty array?

Reply:
it would be null from the `_`, but can't an iterator instance be re-used? the init method gets called before every scan, so if e.g. the sft changes the null bytes would need to be re-calculated